core/src/main/scala/com/raphtory/internals/components/repositories/LocalServiceRepository.scala

```diff
@@ -0,0 +1,30 @@
+package com.raphtory.internals.components.repositories
+
+import cats.effect.Async
+import cats.effect.Ref
+import cats.effect.Resource
+import cats.syntax.all._
+import com.raphtory.internals.communication.TopicRepository
+import com.raphtory.internals.components.ServiceDescriptor
+import com.raphtory.internals.components.ServiceRepository
+
+class LocalServiceRepository[F[_]: Async](topics: TopicRepository, services: Ref[F, Map[String, Any]])
+        extends ServiceRepository[F](topics) {
+
+  override def registered[T](instance: T, descriptor: ServiceDescriptor[F, T]): Resource[F, Unit] =
+    Resource.make(services.update(services => services + (descriptor.name -> instance)))(_ =>
+      services.update(services => services - descriptor.name) // TODO: this doesn't work for multiple instances
+    )
+
+  override def getService[T](descriptor: ServiceDescriptor[F, T]): Resource[F, T] =
+    Resource.eval(services.get.map(serviceList => serviceList(descriptor.name).asInstanceOf[T])) // TODO: this can fail
```

Comment:
no `TODO` returning a wrong instance of a service is a total failure and will be propagated via `F[_]`

Reply:
By that I mean it works as expected